torchvision/datasets/_optical_flow.py

```diff
@@ -0,0 +1,149 @@
+import os
+from abc import ABC, abstractmethod
+from glob import glob
+from pathlib import Path
+
+import numpy as np
+import torch
+from PIL import Image
+
+from ..io.image import _read_png_16
+from .vision import VisionDataset
+
+
+__all__ = (
+    "KittiFlow",
+    "Sintel",
+)
+
+
+class FlowDataset(ABC, VisionDataset):
+    def __init__(self, root, transforms=None):
+
+        super().__init__(root=root)
+        self.transforms = transforms
+
+        self._flow_list = []
+        self._image_list = []
+
+    def _read_img(self, file_name):
+        return Image.open(file_name)
+
+    @abstractmethod
+    def _read_flow(self, file_name):
+        # Return the flow or a tuple (flow, valid) for datasets where the valid mask is built-in
+        pass
+
+    def __getitem__(self, index):
+        # Some datasets like Kitti have a built-in valid mask, indicating which flow values are valid
+        # For those we return (img1, img2, flow, valid), and for the rest we return (img1, img2, flow),
+        # and it's up to whatever consumes the dataset to decide what `valid` should be.
```

Comment:
What do we think of this? Essentially this means that one will do

```py
for e in KittiFlow():
    img1, img2, flow, valid = e

for e in Sintel():
    img1, img2, flow = e
    valid = 2_norm_of_flow_here < SOME_BIG_THRESHOLD  # Typically 1000

# See https://github.com/NicolasHug/vision/blob/6bd34ed5dfbf2710a4865e981db0764ba535007c/references/raft/train.py#L208-L212
```

I'm a bit torn here. One one hand it's not convenient that different datasets have a different return size, but on the other hand we don't want to decide what `valid` should be on behalf of the user when `valid` isn't built in the dataset.

Maybe one option would be to always return `img1, img2, flow, valid` and to set `valid` to None. But it would be a bit awkward for Sintel (and for other future datasets like FlyingChairs and FlyingThings3D) to always have a return value that's always None.

Reply:
> I'm a bit torn here. One one hand it's not convenient that different datasets have a different return size

This will be resolved by the new style datasets, because they always return a dictionary, with more or less information with respect to the dataset. So for now I would go with the different return sizes and equalize them in the reference script:

```python
for img1, img2, flow, *other_features in dataset:
    if not other_features:
        valid = flow.abs() < SOME_BIG_THRESHOLD
    else:
        valid = other_features[0]
```

We will probably always need some kind of equalization, because the datasets authors rarely follow a pattern.